File: text_editor.py
"""
Patrón Command - Editor de Texto
Receiver: TextEditor
"""


class TextEditor:
    """Receiver - El editor de texto que ejecuta las operaciones reales."""

    def __init__(self):
        self._text: list[str] = []  # Simula un StringBuilder
        self._clipboard: str = ""

    # ── Operaciones del editor ──────────────────────────────────

    def write(self, position: int, text: str) -> None:
        """Inserta texto en la posición indicada."""
        content = self.get_text()
        new_content = content[:position] + text + content[position:]
        self._text = list(new_content)

    def delete(self, start: int, length: int) -> str:
        """Elimina 'length' caracteres desde 'start'. Retorna el texto eliminado."""
        content = self.get_text()
        deleted = content[start:start + length]
        new_content = content[:start] + content[start + length:]
        self._text = list(new_content)
        return deleted

    def copy(self, start: int, length: int) -> None:
        """Copia 'length' caracteres desde 'start' al clipboard."""
        content = self.get_text()
        self._clipboard = content[start:start + length]

    def cut(self, start: int, length: int) -> str:
        """Corta 'length' caracteres desde 'start' al clipboard. Retorna el texto cortado."""
        content = self.get_text()
        self._clipboard = content[start:start + length]
        cut_text = self._clipboard
        new_content = content[:start] + content[start + length:]
        self._text = list(new_content)
        return cut_text

    def paste(self, position: int) -> None:
        """Pega el contenido del clipboard en la posición indicada."""
        if self._clipboard:
            self.write(position, self._clipboard)

    # ── Getters ─────────────────────────────────────────────────

    def get_text(self) -> str:
        return "".join(self._text)

    def get_clipboard(self) -> str:
        return self._clipboard

    def set_clipboard(self, text: str) -> None:
        self._clipboard = text

File: java/src/main/java/com/command/TextEditor.java
package com.command;

/**
 * Patrón Command - Editor de Texto
 * Receiver: TextEditor
 */
public class TextEditor {

    private String clipboard = "";          // - clipboard: String
    private final StringBuilder text = new StringBuilder();  // - text: StringBuilder

    // ── Operaciones del editor (+ public) ───────────────────────

    /** Inserta texto en la posición indicada. */
    public void write(int position, String text) {
        this.text.insert(position, text);
    }

    /** Elimina 'length' caracteres desde 'start'. Retorna el texto eliminado. */
    public String delete(int start, int length) {
        String deleted = this.text.substring(start, start + length);
        this.text.delete(start, start + length);
        return deleted;
    }

    /** Copia 'length' caracteres desde 'start' al clipboard. */
    public void copy(int start, int length) {
        this.clipboard = this.text.substring(start, start + length);
    }

    /** Corta 'length' caracteres desde 'start' al clipboard. */
    public String cut(int start, int length) {
        this.clipboard = this.text.substring(start, start + length);
        String cutText = this.clipboard;
        this.text.delete(start, start + length);
        return cutText;
    }

    /** Pega el contenido del clipboard en la posición indicada. */
    public void paste(int position) {
        if (!this.clipboard.isEmpty()) {
            write(position, this.clipboard);
        }
    }

    public String getClipboard() {
        return this.clipboard;
    }

    public String getText() {
        return this.text.toString();
    }
}
